## Table of Contents
- [Introduction](#introduction)
- [Data Wrangling](#wrangling)
    - [Gather](#gather)
    - [Import](#import)
    - [Assess](#assess)
    - [Clean](#clean)
    - [Store](#store)

## I) Introduction <a id = "introduction">

**Broad question:** How do price forecasts for each firm in the S&P 2019 Index compare to their corresponding actual prices?

**Approach:** Analyze difference in means between average forecast EPS and average actual EPS for each firm.

I will be analyzing quarterly price returns within the past 20 years for the firms present in the S&P 500 2019 Index.

> At first, I wanted to analyze the forecasted vs. actual price earnings of the S&P in its entirety for the past 20 years. However, considering that firms continuously enter and leave stock indices every year, there would be varying levels of inconsistencies and marginal errors when comparing annual S&P returns alone. To combat this problem, I have isolated these two approaches:
- Analyze the historical earnings of *only* the firms present in the S&P 2019 Index
- Keep track of all firms that were present in the S&P for the past 20 years. Keep track of how many times each firm appeared in the Index and for those with the least count, analyze them individually on how they differ from the firms that stayed for longer.

[TK] HEre is a breakdown of my final clean CSV's features, TK.csv.

## II) Data Wrangling <a id="wrangling"></a>

To gather the data depicted under the `./data` folder, I used Bloomberg Excel functions.

### A) Gather <a id = "gather"></a>
> **APPROACH 1:** Focus on the firms that appear in the 2019 S&P Index and analyze their forecasted vs. actual price earnings for the last 20 years.

To ensure consistency in analysis among multiple firms, I divide both the forecasted and actual price earning dates by *calendar period* instead of fiscal period. This is because fiscal period differs by firm whereas calendar period is consistent by dates. 

#### Through the Bloomberg Excel functions, I gathered four datasets with different purposes:

- historical forecasted EPS
- historical actual EPS
- historical actual EOD price
- historical forecasted EPS relying on terms

---
Before delving into the data, let's define what above terms:

**EPS**

> EPS stands for ***Estimated Price Earnings.*** The formal definition of EPS given by Investopedia is this:

Earnings per share is the portion of a company's profit that is allocated to each outstanding share of a common stock, serving as an indicator of the company's financial health.

In other words, the EPS is a portion of the company's **net income** after all of their dividends are paid off. Dividends are profits that are paid out to shareholders of the company. EPS is one of the most useful and valuable financial measurements because they ***determine a stock's worth.*** The higher the stock, the more the company can pay out dividends to its shareholders, and the more net profit they are determined to generate.

$$ EPS = \frac{Net Income - Preferred Dividends}{Weighted Average Common Shares Outstanding}\$$

**EOD**

> EOD stands for the ***End of Day*** price. For any given day, the EOD marks the ***price at which the stock was valued*** at the end of the day's trading period.

In [1]:
PATH = './data/'
PATH_CLEAN = './data/clean/'

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

### C) Importing the Data <a id = "import"></a>

### Let's summarize the contents of the following DataFrames as we import them.
> All DataFrames consist of the 505 firms found in the 2019 S&P Index with EPS and EOD data encompassing 20 years: from January 1999 until the December 2019.

**Historic forecasted EPS**
> According to Investopedia, consensus estimates is normally an average or median of all the forecasts from individual analysts tracking a particular stock. In this case, the consensus estimate is for ***EPS for each firm present in the index as of 2019.*** Forecasted EPS is calculated by ***quarterly earnings,*** usually by each firm's fiscal period. Estimates of quarterly earnings are published at the beginning of each quarterly period.

In [4]:
#historic forecasted EPS 
df_eps_fc = pd.read_csv(PATH + 'sp-eps-fc.csv')

In [5]:
df_eps_fc.head()

Term Forecasted  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
0          1999Q1          NaN            NaN            NaN           0.025   
1          1999Q2          NaN            NaN            NaN           0.020   
2          1999Q3          NaN            NaN            NaN           0.023   
3          1999Q4          NaN            NaN            NaN           0.016   
4            00Q1          NaN            NaN            NaN           0.032   

   ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  \
0             NaN            NaN           -0.10          0.423   
1             NaN            NaN           -0.13          0.420   
2             NaN            NaN           -0.11          0.381   
3             NaN            NaN             NaN          0.432   
4             NaN            NaN             NaN          0.441   

   ACN UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
0            NaN  ...          0.456           0.090          0.239   
1            NaN  ...          0.213           0.090          0.264   
2            NaN  ...          0.767           0.099          0.300   
3            NaN  ...          0.432           0.124          0.386   
4            NaN  ...          0.283           0.143          0.436   

   XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  ZBH UN Equity  \
0           0.123            NaN            NaN          0.123            NaN   
1           0.135            NaN            NaN          0.150            NaN   
2           0.129            NaN            NaN          0.171            NaN   
3           0.178            NaN            NaN          0.164            NaN   
4           0.138            NaN            NaN          0.138            NaN   

   ZION UW Equity  ZTS UN Equity  
0           0.622            NaN  
1           0.661            NaN  
2           0.693            NaN  
3           0.667            NaN  
4           0.668            NaN  

[5 rows x 506 columns]

**Historic actual EPS**
> Unlike forecasted EPS, actual EPS are the real numbers denoting Earnings-per-Share for a singular firm. Historic actual EPS will be compared to forecasted EPS to draw correlations and comparisons.

In [6]:
#historic actual EPS
df_eps_act = pd.read_csv(PATH + 'sp-eps-act.csv')

In [7]:
df_eps_act.head()

Quarter  Year  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
0      Q1  1999         0.16           0.99            NaN        0.040000   
1      Q2  1999         0.35           1.76            NaN        0.035357   
2      Q3  1999         0.30           1.86            NaN        0.050357   
3      Q4  1999         0.32           1.89            NaN        0.024643   
4      Q1  2000         0.29           0.89            NaN        0.040585   

   ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  ...  \
0             NaN         0.0925          -0.030           0.44  ...   
1             NaN         0.1000          -0.145           0.42  ...   
2             NaN         0.1075          -0.135           0.30  ...   
3             NaN         0.0375          -0.080           0.43  ...   
4             NaN         0.1050          -0.045           0.45  ...   

   XEL UW Equity  XLNX UW Equity  XOM UN Equity  XRAY UW Equity  \
0           0.34          0.0000          0.210        0.123333   
1           0.06          0.0975          0.285        0.133333   
2           0.63          0.0325          0.315        0.130000   
3           0.43          0.1300          0.600        0.180000   
4           0.45          0.1650          0.500        0.140000   

   XRX UN Equity  XYL UN Equity  YUM UN Equity  ZBH UN Equity  ZION UW Equity  \
0           1.64            NaN         0.1725            NaN            0.56   
1           2.28            NaN         0.2900            NaN            0.60   
2           1.96            NaN         0.3200            NaN            0.64   
3           1.72            NaN         0.2375            NaN            0.49   
4          -1.48            NaN         0.2025            NaN           -0.33   

   ZTS UN Equity  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  

[5 rows x 507 columns]

**Historic actual EOD**
> Though this is not directly related to EPS data, EOD would be an interesting measure to use when generating intriguing visualizations and analyses. Who knows what visuals and conclusions I would arrive to with this measure. 

In [8]:
#historic actual EOD
df_eod_act = pd.read_csv(PATH + 'sp-eod-act.csv')

In [9]:
df_eod_act.head()

date  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
0   3/31/1999          NaN            NaN            NaN          1.2835   
1   6/30/1999          NaN            NaN            NaN          1.6540   
2   9/30/1999          NaN            NaN            NaN          2.2612   
3  12/31/1999      52.0909            NaN            NaN          3.6719   
4   3/31/2000      70.0721            NaN            NaN          4.8504   

   ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  \
0             NaN         8.2934           6.250        20.9504   
1             NaN         6.1859           6.875        20.3630   
2             NaN         5.7462           7.750        16.4471   
3             NaN         3.6843          18.375        16.2513   
4             NaN         3.6388          20.250        15.7478   

   ACN UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
0            NaN  ...            NaN         20.2813        35.2813   
1            NaN  ...            NaN         28.6250        38.5625   
2            NaN  ...            NaN         32.7656        37.9688   
3            NaN  ...            NaN         45.4688        40.2813   
4            NaN  ...            NaN         82.8125        38.9063   

   XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  ZBH UN Equity  \
0          7.7500       140.6214            NaN        12.6284            NaN   
1          9.6250       155.6057            NaN         9.7297            NaN   
2          7.5833       110.4883            NaN         7.3591            NaN   
3          7.8750        59.7723            NaN         6.9434            NaN   
4          9.4583        68.4994            NaN         5.5839            NaN   

   ZION UW Equity  ZTS UN Equity  
0         66.5000            NaN  
1         63.5000            NaN  
2         55.1250            NaN  
3         59.1875            NaN  
4         41.6250            NaN  

[5 rows x 506 columns]

**Historic forecasted EPS 3 months prior**
> Instead of using forecast data collected at the beginning of the fiscal period, this feature contains EPS data projected 3 months before the current fiscal period. This is an interesting metric to see how differently forecasters make their predictions at different times.

In [10]:
#historic forecasted EPS 3-months prior
df_eps_fc_terms = pd.read_csv(PATH + 'sp-eps-fc-terms.csv')

In [11]:
df_eps_fc_terms.head()

Forecast Made Term Forecasted  A UN Equity  AAL UW Equity  AAP UN Equity  \
0     10/1/1999            00Q1          NaN            NaN            NaN   
1      1/1/2000            00Q2          NaN            NaN            NaN   
2      4/1/2000            00Q3          NaN            NaN            NaN   
3      7/1/2000            00Q4          NaN            NaN            NaN   
4      9/1/2000            01Q1          NaN            NaN            NaN   

   AAPL UW Equity  ABBV UN Equity  ABC UN Equity  ABMD UW Equity  \
0           0.030             NaN            NaN             NaN   
1           0.026             NaN            NaN             NaN   
2           0.029             NaN            NaN             NaN   
3           0.032             NaN            NaN           -0.18   
4           0.041             NaN            NaN           -0.09   

   ABT UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
0          0.466  ...            NaN           0.143          0.398   
1          0.437  ...            NaN           0.168          0.378   
2          0.414  ...           0.84           0.189          0.425   
3          0.477  ...           0.53           0.211          0.487   
4          0.495  ...           0.40           0.259          0.475   

   XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  ZBH UN Equity  \
0           0.140            NaN            NaN          0.144            NaN   
1           0.153            NaN            NaN          0.184            NaN   
2           0.149            NaN            NaN          0.218            NaN   
3           0.200            NaN            NaN          0.244            NaN   
4           0.158            NaN            NaN          0.178            NaN   

   ZION UW Equity  ZTS UN Equity  
0           0.777            NaN  
1           0.790            NaN  
2           0.803            NaN  
3           0.777            NaN  
4           0.750            NaN  

[5 rows x 507 columns]

## B) Assess <a id = "assess"></a>

> The following DataFrames contain data for each firm across various dates. To account for all firm averages, my goal is to generate a CSV file where each row contains the firm average, with the features as columns.

In [12]:
dict_dfs = {'eps_fcast' : df_eps_fc,
           'eps_actual' : df_eps_act,
           'eod_actual' : df_eod_act,
           'eps_fcast_terms' : df_eps_fc_terms}

### Check Missing Data

In [13]:
for key, df in dict_dfs.items():
    print(key, df.shape)

eps_fcast (84, 506)
eps_actual (84, 507)
eod_actual (84, 506)
eps_fcast_terms (80, 507)


***We need to make sure the number of firms in each DataFrame is consistent.***

---

**Observation 1:** for `eps_fcast`, there are 505 firms encompassing 84 quarterly fiscal periods since 1999.

> There are 506 columns: 1 column being `Term Forecast`, the rest firm names.

**Observation 2:** for `eps_actual`, there are 505 firms encompassing 84 quarterly fiscal periods since 1999.

> There are 507 columns: 2 columns being `Quarter` and `Year`, the rest firm names.

**Observation 3:** for `eod_actual`, there are 505 firms encompassing 84 quarterly calendar periods since 1999.

> There are 506 columns: 1 column being `date`, the rest firm names.

**Observation 4:** for `eps_fcast_terms`, there are 505 firms encompassing 80 quarterly calendar periods since 1999.

> There are 507 columns: 2 columns being `Forecast Made` and `Term Forecasted`, the rest firm names.
 
Since there are only 80 quarterly calendar periods, that ***implies an entire year is missing.***

**Observation 5:** For `eps_fcast`, `eps_actual`, and `eod_actual`, since there are 4 quarters in a year, 84 quarterly forecast periods equate to 21 years. This is correct since we are analyzing the years from 1999 until the end of 2019.

### Most importantly, the number of firms across all DataFrames is consistent.

----

***Check which year is missing from `eps_fcast_terms`***.

> Each **Term Forecasted** entry under `eps_fcast_terms` records the year with 2 digits, so Quarter 1 of the year 2000 becomes 00Q1.
- isolate the first 2 characters to get the year
- join a '20' in front of the string so 00 becomes 2000
- list the number of unique values.

In [14]:
#iterate over all years in eps_fcast_terms, append '20' in front of string
list(map(lambda x: '20' + x, df_eps_fc_terms['Term Forecasted'].str[:2].unique()))

['2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013',
 '2014',
 '2015',
 '2016',
 '2017',
 '2018',
 '2019']

**Observation**: The year 1999 is missing from `df_eps_fc_terms`. This makes sense because the start of the forecasting period would be in the last quarter of 1999, which is October. 

---

### Check Nulls

In [15]:
for key, df in dict_dfs.items():
    #display (row, column) per DataFrame
    print('# NaN in {}: {}'.format(key, df.isna().sum().sum()))

# NaN in eps_fcast: 7055
# NaN in eps_actual: 5021
# NaN in eod_actual: 6921
# NaN in eps_fcast_terms: 7080


**Observation:** All four DataSets contain null values.
> In order to combat this, we'll have to look at both the **number of rows** and **number of columns** with missing data, separately. This way, we can isolate which firms and/or time periods contain complete or incomplete data.

In [16]:
#check rows for missing data
for key, df in dict_dfs.items():
    num_rows_missing = df.isna().sum().max()
    print('{} has {} time periods containing missing data, out of {} total rows.'.format(key, num_rows_missing, df.shape[0]))

eps_fcast has 82 time periods containing missing data, out of 84 total rows.
eps_actual has 84 time periods containing missing data, out of 84 total rows.
eod_actual has 83 time periods containing missing data, out of 84 total rows.
eps_fcast_terms has 80 time periods containing missing data, out of 80 total rows.


In [17]:
len(df_eps_fc.columns[df_eps_fc.isnull().any()])

254

In [18]:
#check columns for missing data
for key, df in dict_dfs.items():
    cols_missing = df.columns[df.isnull().any()]
    num_cols_missing = len(cols_missing)
    print('{} has {} firms containing missing data, out of {} total columns.'
         .format(key, num_cols_missing, df.shape[1]))

eps_fcast has 254 firms containing missing data, out of 506 total columns.
eps_actual has 432 firms containing missing data, out of 507 total columns.
eod_actual has 505 firms containing missing data, out of 506 total columns.
eps_fcast_terms has 340 firms containing missing data, out of 507 total columns.


**Observation 1:** The only two datasets with incomplete data for all time periods are **actual EPS** and **forecasted EPS 3 months prior.**

> To address this problem, it'd be helpful to isolate the time period ranges for the datasets with incomplete data by row, **forecasted EPS** and **actual EOD price.**

**Observation 2:** For all datasets, all firms contain incomplete data across all time periods.
> This is expected, as analyzing financial history spanning over 20 years will naturally be rife with missing and inaccurate data. The ***good news is that `eps_fcast`, `eps_actual`, and `eps_fcast_terms` are the most complete, while `eod_actual` contains the most amount of missing data.***


**Moving forward, we need to make sure that these inconsistencies won't clash with our analysis.**
> **My approach:** instead of looking at rows and columns ***with*** missing data, we'll be looking at rows and columns that ***are all missing data.***

I figured that if there is some missing data here and there scattered throughout the matrix, then that should not skew our analysis too much.

However, if there a significant amount of rows/columns that are entirely empty, then we ***might have to get ready to drop some dates and firms from our data overall.***

In [19]:
#check for empty rows, return False if row contains at least one non-null value, True if all are null
for key, df in dict_dfs.items():
    cols_check = df.columns
    num_empty_rows = (df[cols_check].isnull().apply(lambda x: all(x), axis = 1)).value_counts()
    print(key, '\n', num_empty_rows, '\n----')

eps_fcast 
 False    84
dtype: int64 
----
eps_actual 
 False    84
dtype: int64 
----
eod_actual 
 False    84
dtype: int64 
----
eps_fcast_terms 
 False    80
dtype: int64 
----


**Observation:** All datasets do not contain empty rows.
> This is good news, since we can rely on the firms' averages per row instead of having to drop or limit time periods.

In [20]:
#check for empty columns, return False if column contains at least one non-null value, True if all are null
for key, df in dict_dfs.items():
    cols_check = df.columns
    num_empty_cols = df[cols_check].isnull().apply(lambda x: all(x), axis = 0).value_counts()
    print(key, '\n', num_empty_cols)

eps_fcast 
 False    506
dtype: int64
eps_actual 
 False    506
True       1
dtype: int64
eod_actual 
 False    506
dtype: int64
eps_fcast_terms 
 False    506
True       1
dtype: int64


**Observation:** `eps_actual` and `eps_fcast_terms` are the only datasets that have an empty column.
> Let's isolate and look at the singular empty column for both DataFrames.

In [21]:
#create function to return an array of column names containing empty data
def comb_cols(df):
    empty_cols = []
    for column in df:
        if df[column].isnull().all():
            empty_cols.append(column)
            
    return empty_cols

In [22]:
#comb datasets for empty columns
print('In eps_act, the firm {} has no data.'.format(comb_cols(df_eps_act)))
print('In eps_fc_terms, the firm {} has no data.'.format(comb_cols(df_eps_fc_terms)))

In eps_act, the firm ['AMCR UN Equity'] has no data.
In eps_fc_terms, the firm ['AMCR UN Equity'] has no data.


**Observation:** The same firm in both datasets is empty of data.
> Though this is an annoying error to deal with, it still is to our advantage that both datasets ***share one firm*** in common for missing data. This way, we don't have to worry about dropping two entire firms.

---

### Check Duplicate Data

In [23]:
#check for duplicate data across all rows and columns
for key, df in dict_dfs.items():
    print(key, df.duplicated().sum())

eps_fcast 0
eps_actual 0
eod_actual 0
eps_fcast_terms 0


**Observation:** For all datasets, there are ***no duplicate data.*** This is good news!

**Next, I will check for duplicated firm names.** Although the presence of duplicated firm names will inherently imply duplicated data, sometimes data gets dispersed in weird, unexpected ways, especially when dealing with large datasets.

In [24]:
#check for duplicated firm names
for key, df in dict_dfs.items():
    print(key, df.columns.duplicated().sum())

eps_fcast 0
eps_actual 0
eod_actual 0
eps_fcast_terms 0


**Observation:** For all datasets, there are ***no duplicate firm names.*** This is also good news.

### There is no need to dedupe our data during the cleaning stage.

---

### Check Data Types

> Make sure all numerical data types are consistent.
- EPS and EOD values must all be **floats**
- Dates should all be **objects** (for now)

- `df_eps_fc` contains ***1 date field,***, so we should expect ***1 object type.***
- `df_eps_act` contains ***2 date fields,*** so we should expect ***2 object types.***
- `df_eod_act` contains ***1 date field,*** so we should expect ***1 object type.***
- `df_eps_fc_terms` contains ***2 date fields,*** so we should expect ***2 object types.***

> And everything else should be a ***float*** type.

**Examine each dataset's overall data types per column.**

In [25]:
for key, df in dict_dfs.items():
    print(key,'\n', df.dtypes.value_counts(), '\n--------')

eps_fcast 
 float64    505
object       1
dtype: int64 
--------
eps_actual 
 float64    499
object       7
int64        1
dtype: int64 
--------
eod_actual 
 float64    505
object       1
dtype: int64 
--------
eps_fcast_terms 
 float64    505
object       2
dtype: int64 
--------


**Observation 1:** `df_eps_fc`, `df_eod_act`, and `df_eps_fc_terms` are all consistent with the expected number of object types.

> But we'll still have to double check these facts later.

**Observation 2:** `df_eps_act` contains 5 more object columns than expected.

---

***Isolate 'Object' Columns under `df_eps_act`***

In [26]:
df_eps_act.select_dtypes(include = 'object').head()

Quarter            BRK/B UN Equity              FOX UW Equity  \
0      Q1  #N/A Field Not Applicable  #N/A Field Not Applicable   
1      Q2  #N/A Field Not Applicable  #N/A Field Not Applicable   
2      Q3  #N/A Field Not Applicable  #N/A Field Not Applicable   
3      Q4  #N/A Field Not Applicable  #N/A Field Not Applicable   
4      Q1  #N/A Field Not Applicable  #N/A Field Not Applicable   

              GOOG UW Equity              HCP UN Equity  \
0  #N/A Field Not Applicable  #N/A Field Not Applicable   
1  #N/A Field Not Applicable  #N/A Field Not Applicable   
2  #N/A Field Not Applicable  #N/A Field Not Applicable   
3  #N/A Field Not Applicable  #N/A Field Not Applicable   
4  #N/A Field Not Applicable  #N/A Field Not Applicable   

              SYMC UW Equity               UA UN Equity  
0  #N/A Field Not Applicable  #N/A Field Not Applicable  
1  #N/A Field Not Applicable  #N/A Field Not Applicable  
2  #N/A Field Not Applicable  #N/A Field Not Applicable  
3  #N/A Field Not Applicable  #N/A Field Not Applicable  
4  #N/A Field Not Applicable  #N/A Field Not Applicable

**Observation:** These firms are all 'Object' types because there is no recorded data under them.
> This is actually missing data. From `df_eps_act`, there are 6 firms that are empty of data.

### This was a tricky situation to spot, but I caught it in the end. There actually is missing data after all, but we missed those 5 entire firms during the "Check Missing Data" stage, because they were recorded as objects. This means that Pandas wrongly recognized these 5 firms as "complete."

### Quality

**Missing Data**
> **Deal with dropping/keeping missing data in the end when all data is properly cleaned.**

-  `df_eps_fc_terms` is missing the year 1999.
- `df_eps_act` and `df_eps_fc_act` have one firm with empty data: 'AMCR UN Equity'
--- 
- Unnormalized date formats among all DataFrames.
- Firm names across all DataFrames are capitalized and contain white space.
- Erroneous data type for `df_eps_act` Object columns.
- Erroneous data type for **date** under `df_eod_act`
- Erroneous data type for **forecast_made** under `df_eps_fc_terms`
- Turn YYYYQQ into datetime quarter format?? (research)
- No recorded 20-year for each dataset.
- No recorded yearly averages for each dataset
- No recorded quarterly averages for each dataset.

### Tidiness
- Firm 20-year, yearly, and quarterly averages contained in different DataFrames.
- forecast + actual EPS DataFrames need to be merged with firm names transposed into rows


## C) CLEAN



In [275]:
eps_act_clean

term_forecast     a       aal       aap      aapl      abbv       abc  \
0         1999Q1  0.16  0.990000       NaN  0.040000       NaN  0.092500   
1         1999Q2  0.35  1.760000       NaN  0.035357       NaN  0.100000   
2         1999Q3  0.30  1.860000       NaN  0.050357       NaN  0.107500   
3         1999Q4  0.32  1.890000       NaN  0.024643       NaN  0.037500   
4         2000Q1  0.29  0.890000       NaN  0.040585       NaN  0.105000   
5         2000Q2  0.37  2.150000       NaN  0.051429       NaN  0.117500   
6         2000Q3  0.34  2.080000       NaN  0.044286       NaN  0.122500   
7         2000Q4  0.67  0.310000       NaN  0.037143       NaN  0.135000   
8         2001Q1  0.22 -0.280000  0.046667 -0.041429       NaN  0.125000   
9         2001Q2  0.22 -3.290000  0.166667  0.008571       NaN  0.150000   
10        2001Q3 -0.48 -2.680000  0.180000  0.012143       NaN  0.150000   
11        2001Q4  0.43 -5.170000 -0.243333  0.013571       NaN  0.115000   
12        2002Q1 -0.68 -3.710000  0.120000  0.007857       NaN  0.162500   
13        2002Q2 -0.55 -3.190000  0.150000  0.007857       NaN  0.220000   
14        2002Q3 -0.49 -5.920000  0.266667  0.006429       NaN  0.215000   
15        2002Q4 -0.51 -3.390000       NaN -0.009286       NaN  0.222500   
16        2003Q1 -0.78 -6.680000  0.046667 -0.001429       NaN  0.217500   
17        2003Q2 -0.31 -0.470000  0.400000  0.002857       NaN  0.265000   
18        2003Q3 -3.28  0.000000  0.406667  0.003571       NaN  0.255000   
19        2003Q4  0.03 -0.700000  0.280000  0.008571       NaN  0.267500   
20        2004Q1  0.15 -1.030000  0.460000  0.012143       NaN  0.242500   
21        2004Q2  0.22  0.040000  0.473333  0.009286       NaN  0.317500   
22        2004Q3  0.21 -1.330000  0.306667  0.011429       NaN  0.280000   
23        2004Q4  0.19 -2.400000  0.293333  0.020000       NaN  0.207500   
24        2005Q1  0.21 -1.000000       NaN  0.053571       NaN  0.060000   
25        2005Q2  0.19  0.350000  0.606667  0.051429       NaN  0.225000   
26        2005Q3   NaN -0.930000       NaN  0.055714       NaN  0.227500   
27        2005Q4  0.05 -3.460000  0.360000  0.074286       NaN  0.047500   
28        2006Q1  5.95 -0.490000  0.690000  0.097143       NaN  0.235000   
29        2006Q2  0.27  1.440000  0.600000  0.070000       NaN  0.310000   
..           ...   ...       ...       ...       ...       ...       ...   
54        2012Q3  0.70 -0.710000  1.220000  1.345714  1.010000  0.720000   
55        2012Q4  1.22  0.785075  0.890000  1.251429  0.976538  0.670000   
56        2013Q1  0.52 -1.370000  1.660000  1.990000  0.610000  0.730000   
57        2013Q2  0.48  0.880000  1.600000  1.451429  0.670000  0.200000   
58        2013Q3  0.50  1.160000  1.420000  1.072857  0.600000  0.730000   
59        2013Q4  0.64 -5.380000  0.680000  1.187143  0.700000  0.220000   
60        2014Q1  0.59  0.660000  2.020000  2.084286  0.610000  0.180000   
61        2014Q2  0.42  1.200000  1.910000  1.670000  0.690000  0.780000   
62        2014Q3  0.44  1.310000  1.670000  1.290000  0.320000 -0.060000   
63        2014Q4  0.07  0.840000  1.150000  1.430000 -0.510000  0.300000   
64        2015Q1  0.19  1.340000  2.020000  3.080000  0.640000 -0.910000   
65        2015Q2  0.26  2.470000  2.040000  2.340000  0.840000 -2.330000   
66        2015Q3  0.33  2.560000  1.640000  1.860000  0.750000  0.980000   
67        2015Q4  0.42  5.240000  0.750000  1.970000  0.930000  1.710000   
68        2016Q1  0.37  1.150000  2.160000  3.300000  0.830000  1.600000   
69        2016Q2  0.28  1.690000  1.690000  1.910000  0.990000  2.900000   
70        2016Q3  0.38  1.400000  1.540000  1.430000  0.970000  1.620000   
71        2016Q4  0.38  0.560000  0.840000  1.680000  0.850000  0.660000   
72        2017Q1  0.52  0.670000  1.460000  3.380000  1.070000  1.130000   
73        2017Q2  0.51  1.760000  1.180000  2.110000  1.200000  1.890000   
74        2017Q3  0.55  1.360000  1.300000  1.68000

In [200]:
eps_fc_clean = df_eps_fc.copy()
eps_act_clean = df_eps_act.copy()
eod_act_clean = df_eod_act.copy()
eps_fc_terms_clean = df_eps_fc_terms.copy()

In [201]:
#create dictionary of all clean dfs for iteration
dict_clean = {'eps_fc_clean' : eps_fc_clean,
             'eps_act_clean' : eps_act_clean,
             'eod_act_clean' : eod_act_clean,
             'eps_fc_terms_clean' : eps_fc_terms_clean}

### Code
> **ISSUE 1:**  Unnormalized date formats among all DataFrames.

**Define:** 
- Conjoin `eps_act_clean` dates from 2 columns into 1 to match `eps_fc_clean` format.
> The format we want is YYYYQN. For example, Quarter 1 in 2005 will be 2005Q1.

- Rename both fiscal periods under `eps_act_clean` and `eps_fc_clean` to **term_forecast.**

- For both fiscal periods under `eps_fc_terms_clean`, add an underscore.
- Under `eps_fc_terms_clean`, format **term_forecasted** to YYYYQN.

In [203]:
#create new column, combine year and quarter into 1 string
eps_act_clean['term_forecast'] = eps_act_clean['Year'].map(str) + eps_act_clean['Quarter']

In [204]:
#drop 'Year' and 'Quarter'
eps_act_clean.drop(['Year', 'Quarter'], axis = 1, inplace = True)

In [205]:
#designate new column as first column
reorder_cols = eps_act_clean.columns.tolist()
reorder_cols.insert(0, reorder_cols.pop(reorder_cols.index('term_forecast')))

eps_act_clean = eps_act_clean.reindex(columns = reorder_cols)

In [206]:
#rename eps_fc_clean fiscal period to term_forecast
eps_fc_clean.rename(index = str, columns = {'Term Forecasted' : 'term_forecast'}, inplace = True)

#reassign to dictionary
dict_clean['eps_act_clean'] = eps_act_clean

In [207]:
eps_fc_clean[4:].term_forecast = '20' + eps_fc_clean[4:].term_forecast.astype(str)

In [208]:
#add an underscore for both date periods under eps_fc_terms_clean
eps_fc_terms_clean.rename(index = str, columns = {'Forecast Made' : 'forecast_made',
                                                 'Term Forecasted' : 'term_forecast'},
                         inplace = True)

In [209]:
#under term_forecaste, add a '20' before each string
eps_fc_terms_clean.term_forecast = '20' + eps_fc_terms_clean.term_forecast.astype(str)

### Test

In [210]:
#give overview of both DataFrames
eps_fc_clean.sample(3)

term_forecast  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
71        2016Q4        0.521          0.919          1.085           1.656   
32        2007Q1        0.348          0.303          0.706           0.113   
75        2017Q4        0.624          0.919          0.634           1.870   

    ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  \
71           1.197          1.228           0.190          0.646   
32             NaN          0.279          -0.168          0.535   
75           1.447          1.321           0.360          0.732   

    ACN UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
71          1.300  ...          0.437           0.520          0.704   
32          0.415  ...          0.360           0.298          1.542   
75          1.463  ...          0.426           0.543          1.030   

    XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  \
71           0.650          1.004          0.649          0.734   
32           0.358          0.840            NaN          0.318   
75           0.812          0.961          0.761          0.795   

    ZBH UN Equity  ZION UW Equity  ZTS UN Equity  
71          2.136           0.521          0.453  
32          0.933           1.418            NaN  
75          2.098           0.732          0.658  

[3 rows x 506 columns]

In [211]:
eps_act_clean.sample(3)

term_forecast  A UN Equity  AAL UW Equity  AAP UN Equity  AAPL UW Equity  \
76        2018Q1        -0.99           0.34           1.85        3.920000   
28        2006Q1         5.95          -0.49           0.69        0.097143   
20        2004Q1         0.15          -1.03           0.46        0.012143   

    ABBV UN Equity  ABC UN Equity  ABMD UW Equity  ABT UN Equity  \
76            1.74         3.9500            0.85           0.24   
28             NaN         0.2350           -0.77           0.57   
20             NaN         0.2425           -0.15           0.53   

    ACN UN Equity  ...  XEL UW Equity  XLNX UW Equity  XOM UN Equity  \
76           1.82  ...           0.57            0.63           1.09   
28           0.37  ...           0.37            0.22           1.38   
20           0.34  ...           0.37            0.14           0.83   

    XRAY UW Equity  XRX UN Equity  XYL UN Equity  YUM UN Equity  \
76           0.360           0.08           0.44          1.300   
28           0.315           0.80            NaN          0.310   
20           0.555           1.12            NaN          0.245   

    ZBH UN Equity  ZION UW Equity  ZTS UN Equity  
76           0.86            1.16           0.72  
28           0.83            1.30            NaN  
20           0.40            1.11            NaN  

[3 rows x 506 columns]

In [212]:
#check consistent format between both `term_forecast` fields
forecasts = pd.concat([eps_fc_clean.term_forecast, eps_act_clean.term_forecast])

#count number of instances where unique term_forecast values are the same
(forecasts.value_counts().sort_index() == 2).value_counts()

True    84
Name: term_forecast, dtype: int64

In [213]:
#check that dates are the first columns
for key, df in dict_clean.items():
    print(df.columns[:2])

Index(['term_forecast', 'A UN Equity'], dtype='object')
Index(['term_forecast', 'A UN Equity'], dtype='object')
Index(['date', 'A UN Equity'], dtype='object')
Index(['forecast_made', 'term_forecast'], dtype='object')


## The **term_forecast** fields under `eps_fc_clean`, `eps_act_clean`, and `eps_fc_terms_clean` are all normalized to YYYQN format. This means `eod_act_clean` is the only one with a different date formatting rule.

### Code
> **ISSUE 2:** Firm names across all DataFrames are capitalized and contain white space.

**Define:** Iterate across all DataFrames. Get rid of everything from the first whitespace character onward. Lowercase column names.

In [214]:
#lowercase names
for key, df in dict_clean.items():
    lower_cols = [x.lower() for x in df.columns]
    df.columns = lower_cols

In [215]:
#get rid of white space
for key, df in dict_clean.items():
    stock_symbols = [x.split(' ', 1)[0] for x in df.columns]
    df.columns = stock_symbols

### Test

In [216]:
#check that all firm names are consistent, should return 505 firm results
all_ticks = []
for key, df in dict_clean.items():
    firm_ticks = np.array(df.columns.values)
    all_ticks.append(firm_ticks)
    
#flatten array
all_ticks = np.hstack(all_ticks)

#assign tick names to Pandas Series
all_ticks = pd.Series(all_ticks)

In [217]:
#look at lowest 3 value counts, SHOULD BE DATE PERIODS
all_ticks.value_counts().sort_values()[:5]

date             1
forecast_made    1
term_forecast    3
ntap             4
pnc              4
dtype: int64

In [218]:
#look at highest value counts, should be FOUR (because there are 4 dataframes all with firm names)
all_ticks.value_counts().sort_values(ascending = False)[:5]

fti     4
hii     4
pypl    4
eqix    4
sbux    4
dtype: int64

In [219]:
#check for any upper case instance among columns
all_ticks.str.isupper().any()

False

In [221]:
#check for any white space
all_ticks.str.contains(' ').any()

False

### Code
> **Issue 3:** Erroneous data type for `df_eps_act` Object columns.

**Define:**

- Convert '#N/A Field Not Applicable' strings into 'NaN'
- Convert the 6 `eps_act_clean` Object columns into NaN type.

In [222]:
#isolate 6 firm names recorded as 'object'
obj_ticks = eps_act_clean.select_dtypes(include = 'object').columns.values

#exclude 'term_forecast'
obj_ticks = obj_ticks[1:]

In [223]:
#convert to NaN
str_replace = '#N/A Field Not Applicable'
eps_act_clean[obj_ticks] = eps_act_clean[obj_ticks].replace(str_replace, np.nan)

### Test

In [224]:
eps_act_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84 entries, 0 to 83
Columns: 506 entries, term_forecast to zts
dtypes: float64(505), object(1)
memory usage: 332.1+ KB


In [225]:
#make sure 'term_forecast' is the only object-type column
eps_act_clean.select_dtypes(include = 'object').columns.values.tolist()

['term_forecast']

### Code

> Issue 3a:  Erroneous data type for **date** under `df_eod_act`

**Define:** Convert column `date` to DateTime object.

In [226]:
eod_act_clean.date = pd.to_datetime(eod_act_clean.date, infer_datetime_format=True)

### Test

In [227]:
eod_act_clean.sample(2)

date        a  aal     aap     aapl  abbv      abc   abmd      abt  \
34 2007-09-28  26.3724  NaN  33.560  21.9343   NaN  22.6650  12.43  25.6557   
24 2005-03-31  14.9577  NaN  33.633   5.9529   NaN  13.8977  10.58  22.3064   

      acn  ...  xel   xlnx    xom    xray      xrx  xyl      yum    zbh  \
34  40.25  ...  NaN  26.14  92.56  41.640  45.6838  NaN  24.3256  80.99   
24  24.15  ...  NaN  29.23  59.60  27.205  39.9141  NaN  18.6271  77.81   

     zion  zts  
34  68.67  NaN  
24  69.02  NaN  

[2 rows x 506 columns]

In [228]:
#confirm datetime64 present
eod_act_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84 entries, 0 to 83
Columns: 506 entries, date to zts
dtypes: datetime64[ns](1), float64(505)
memory usage: 332.1 KB


### Code
> **Issue 5:** Erroneous data type for **forecast_made** under `df_eps_fc_terms`.

**Define:**

- Convert column **forecast_made** to DateTime object.

In [229]:
eps_fc_terms_clean.forecast_made = pd.to_datetime(eps_fc_terms_clean.forecast_made, infer_datetime_format=True)

### Test

In [230]:
#confirm presence of datetime64 object
eps_fc_terms_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 80 entries, 0 to 79
Columns: 507 entries, forecast_made to zts
dtypes: datetime64[ns](1), float64(505), object(1)
memory usage: 317.5+ KB


In [231]:
eps_fc_terms_clean.sample(2)

forecast_made term_forecast   a  aal  aap   aapl  abbv  abc  abmd    abt  \
0    1999-10-01        2000Q1 NaN  NaN  NaN  0.030   NaN  NaN   NaN  0.466   
1    2000-01-01        2000Q2 NaN  NaN  NaN  0.026   NaN  NaN   NaN  0.437   

   ...  xel   xlnx    xom   xray  xrx  xyl    yum  zbh   zion  zts  
0  ...  NaN  0.143  0.398  0.140  NaN  NaN  0.144  NaN  0.777  NaN  
1  ...  NaN  0.168  0.378  0.153  NaN  NaN  0.184  NaN  0.790  NaN  

[2 rows x 507 columns]

## All full years are normalized to YYYY-MM-DD, and all quarterly records to YYYY-QQ.

## This is to enable for more efficient handling, cleaning, and classifying of data later on.

### Code
> **Issue 4:** No recorded 20-year averages for each dataset.

**Define:**


- Isolate 20-year averages for each firm into its own DataFrame

- Create new DataFrame `df_twenty_avgs` depicting all 20-year averages for each firm

In [232]:
#grab all firm names
firm_names = eps_act_clean.columns.values[1:]

In [233]:
def get_twenty_yr_avg(df, index_names, col_name):
    return pd.DataFrame({col_name : df[index_names].mean()},
                       index = index_names)

In [234]:
#get 20-year averages
eps_fc_twenty_avg = get_twenty_yr_avg(eps_fc_clean, firm_names, 'eps_fc')
eps_act_twenty_avg = get_twenty_yr_avg(eps_act_clean, firm_names, 'eps_act')
eod_act_twenty_avg = get_twenty_yr_avg(eod_act_clean, firm_names, 'eod_act')
eps_fc_terms_twenty_avg = get_twenty_yr_avg(eps_fc_terms_clean, firm_names, 'eps_fc_terms')

In [235]:
#put averages into a list
twenty_avg_list = [eps_fc_twenty_avg,
                  eps_act_twenty_avg,
                  eod_act_twenty_avg,
                  eps_fc_terms_twenty_avg]

In [236]:
#merge all Series
df_twenty_year_avgs = pd.concat(twenty_avg_list, axis = 1, sort=False)

### Test

In [237]:
df_twenty_year_avgs.sample(10)

eps_fc   eps_act     eod_act  eps_fc_terms
ms    0.736143  0.753434   39.559912      0.958551
bkng  9.167367  5.120181  597.128590     10.053804
apa   1.160590  0.139465   61.038492      1.329675
bll   0.279679  0.217348   19.115859      0.317870
alk   0.786224  0.516265   26.256899      0.886074
hpe   0.387882  0.263333   13.709531      0.411267
wcg   1.042355  1.006389   87.097049      1.216100
ksu   0.804417  0.805802   52.362895      0.888879
lvs   0.478141  0.429333   53.357333      0.533288
tsco  0.376310  0.381128   33.462765      0.430932

In [238]:
#check missing data
df_twenty_year_avgs.isna().any()

eps_fc          False
eps_act          True
eod_act         False
eps_fc_terms     True
dtype: bool

In [239]:
#check duplicate data
df_twenty_year_avgs.duplicated().any()

False

### Code
> **Issue 5:** No recorded yearly averages for each dataset.

**Define:**
- Create 4 separate DataFrames for all attributes
- Rename columns to "Feature_Year" (e.g. eps_fc_1999, eod_act_2000, etc.)
- Outer merge all DataFrames to create new DataFrame, `df_yearly_avgs`, on 

In [265]:
#helper function to calculate yearly average of each DataFrame, then transpose firm names into rows
def create_df_yearly_avgs(df, varX):
    return df.groupby(df[varX].str[:4]).mean().transpose()

In [268]:
eps_fc_yearly_avgs = create_df_yearly_avgs(eps_fc_clean, 'term_forecast')
eps_act_yearly_avgs = create_df_yearly_avgs(eps_act_clean, 'term_forecast')
eod_act_yearly_avgs = create_df_yearly_avgs(eod_act_clean, 'date')

AttributeError: Can only use .str accessor with string values, which use np.object_ dtype in pandas

In [263]:
eod_act_clean

date        a     aal      aap      aapl   abbv       abc     abmd  \
0  1999-03-31      NaN     NaN      NaN    1.2835    NaN    8.2934    6.250   
1  1999-06-30      NaN     NaN      NaN    1.6540    NaN    6.1859    6.875   
2  1999-09-30      NaN     NaN      NaN    2.2612    NaN    5.7462    7.750   
3  1999-12-31  52.0909     NaN      NaN    3.6719    NaN    3.6843   18.375   
4  2000-03-31  70.0721     NaN      NaN    4.8504    NaN    3.6388   20.250   
5  2000-06-30  49.6906     NaN      NaN    3.7411    NaN    7.5201   15.375   
6  2000-09-29  32.9726     NaN      NaN    1.8393    NaN   11.4015   34.063   
7  2000-12-29  36.8889     NaN      NaN    1.0625    NaN   12.2505   24.250   
8  2001-03-30  20.7050     NaN      NaN    1.5764    NaN   11.8988   15.750   
9  2001-06-29  21.8975     NaN      NaN    1.6607    NaN   13.4149   23.580   
10 2001-09-28  13.1722     NaN      NaN    1.1079    NaN   17.2114   17.470   
11 2001-12-31  19.2092     NaN   16.583    1.5643    NaN   15.4162   15.820   
12 2002-03-28  23.5550     NaN   15.333    1.6907    NaN   16.5685   11.100   
13 2002-06-28  15.9347     NaN   18.170    1.2657    NaN   18.4364    8.479   
14 2002-09-30   8.7994     NaN   17.580    1.0357    NaN   17.3254    3.350   
15 2002-12-31  12.1009     NaN   16.300    1.0236    NaN   13.1748    3.640   
16 2003-03-31   8.8601     NaN   15.417    1.0100    NaN   12.7357    3.900   
17 2003-06-30  13.1722     NaN   20.300    1.3657    NaN   16.8232    5.470   
18 2003-09-30  14.8971     NaN   23.633    1.4736    NaN   13.1117    8.610   
19 2003-12-31  19.7010     NaN   27.133    1.5264    NaN   13.6211    6.910   
20 2004-03-31  21.3114     NaN   27.113    1.9321    NaN   13.2645    8.100   
21 2004-06-30  19.7280     NaN   29.453    2.3243    NaN   14.5017   12.580   
22 2004-09-30  14.5332     NaN   22.933    2.7679    NaN   13.0292    8.850   
23 2004-12-31  16.2379     NaN   29.120    4.6000    NaN   14.2349   15.440   
24 2005-03-31  14.9577     NaN   33.633    5.9529    NaN   13.8977   10.580   
25 2005-06-30  15.5102     NaN   43.033    5.2586    NaN   16.7747    8.550   
26 2005-09-30  22.0660     NaN   38.680    7.6586    NaN   18.7518   10.070   
27 2005-12-30  22.4298     NaN   43.460   10.2700    NaN   20.0860    9.240   
28 2006-03-31  25.3001     NaN   41.640    8.9600    NaN   23.4191   12.900   
29 2006-06-30  21.2642     NaN   28.900    8.1600    NaN   20.3383   12.970   
..        ...      ...     ...      ...       ...    ...       ...      ...   
54 2012-09-28  27.4951     NaN   68.440   95.3229    NaN   38.7100   20.990   
55 2012-12-31  29.2757     NaN   72.350   76.1471  34.16   43.1800   13.460   
56 2013-03-28  30.0122     NaN   82.650   63.2329  40.78   51.4500   18.670   
57 2013-06-28  30.5771     NaN   81.170   56.5829  41.34   55.8300   21.560   
58 2013-09-30  36.6482     NaN   82.680   68.1071  44.73   61.1000   19.070   
59 2013-12-31  40.8958  25.250  110.680   80.1586  52.81   70.3100   26.740   
60 2014-03-31  39.9877  36.600  126.500   76.6771  51.40   65.5900   26.040   
61 2014-06-30  41.0746  42.960  134.920   92.9300  56.44   72.6600   25.140   
62 2014-09-30  40.7456  35.480  130.300  100.7500  57.76   77.3000   24.830   
63 2014-12-31  40.9400  53.630  159.280  110.3800  65.44   90.1600   38.060   
64 2015-03-31  41.5500  52.780  149.690  124.4300  58.54  113.6700   71.580   
65 2015-06-30  38.5800  39.935  159.290  125.4250  67.19  106.3400   65.730   
66 2015-09-30  34.3300  38.830  189.530  110.3000  54.41   94.9900   92.760   
67 2015-12-31  41.8100  42.350  150.510  105.2600  59.24  103.7100   90.280   
68 2016-03-31  39.8500  41.010  160.340  108.9900  57.12   86.5500   94.810   
69 2016-06-30  44.3600  28.310  161.630   95.6000  61.91   79.3200  109.290   
70 2016-09-30  47.0900  36.610  149.120  113.0500  63.07   80.7800  128.580   
71 2016-12-30  45.5600  46.690  169.120  115.8200  62.62   78.1900  112.680   
72 2017-03-31  52.8700  42.300  148.260  143.6600  65.16   

In [100]:
#isolate eod_act_yearly_avgs

eod_act_yearly_avgs = eod_act_clean.groupby(eod_act_clean.date.str[-4:]).mean()

#transpose firm names and years
eod_act_yearly_avgs = eod_act_yearly_avgs.transpose()

In [101]:
#isolate eps_fc_terms_yearly_avgs
eps_fc_terms_yearly_avgs = eps_fc_terms_clean.groupby(eps_fc_terms_clean.forecast_made.str[-4:]).mean()

#transpose firm names and years
eps_fc_terms_yearly_avgs = eps_fc_terms_yearly_avgs.transpose()

***CHECK THAT OUR DATA IS CONSISTENT WITH THE ORIGINAL DATA***

***REORDER COLUMNS (IF APPLICABLE)***

# III) Store Data

In [132]:
df_twenty_year_avgs.to_csv(PATH_CLEAN + 'twenty_year_avgs.csv', encoding = 'utf-8', index = False)

# IV) Notes

- I did not normalize between (`eod_act_` and `eps_fc_terms`) and (`eps_fc` and `eps_act`) because the former two are recorded on calendar dates, while the latter two are recorded on fiscal periods.
- When dealing with the CSVs depicting averages, we need to keep in mind **missing data** in our interpretations.


In [ ]:
#convert notebook to HTML